boa/src/builtins/intl/mod.rs

```diff
@@ -0,0 +1,143 @@
+            // a. Let O be CreateArrayFromList(« locales »).
+            Array::create_array_from_list([locales.clone()], context)
+        } else {
+            // 4. Else,
+            // a. Let O be ? ToObject(locales).
+            locales.to_object(context)?
+        };
+
+        // 5. Let len be ? ToLength(? Get(O, "length")).
+        let len = o.length_of_array_like(context)?;
+
+        // 6 Let k be 0.
+        // 7. Repeat, while k < len,
+        for k in 0..len {
+            // a. Let Pk be ToString(k).
+            // b. Let kPresent be ? HasProperty(O, Pk).
+            let k_present = o.has_property(k, context)?;
+            // c. If kPresent is true, then
+            if k_present {
+                // i. Let kValue be ? Get(O, Pk).
+                let k_value = o.get(k, context)?;
+                // ii. If Type(kValue) is not String or Object, throw a TypeError exception.
+                if !(k_value.is_object() || k_value.is_string()) {
+                    return Err(context
+                        .throw_type_error("locale should be a String or Object")
+                        .unwrap_err());
+                }
+                // iii. If Type(kValue) is Object and kValue has an [[InitializedLocale]] internal slot, then
+                // TODO: handle checks for InitializedLocale internal slot (there should be an if statement here)
+                // 1. Let tag be kValue.[[Locale]].
+                // iv. Else,
+                // 1. Let tag be ? ToString(kValue).
+                let tag = k_value.to_string(context)?;
+                // v. If IsStructurallyValidLanguageTag(tag) is false, throw a RangeError exception.
+                // TODO: implement `IsStructurallyValidLanguageTag`
+
+                // vi. Let canonicalizedTag be CanonicalizeUnicodeLocaleId(tag).
+                seen.insert(Self::canonicalize_locale(&tag));
+                // vii. If canonicalizedTag is not an element of seen, append canonicalizedTag as the last element of seen.
+            }
+            // d. Increase k by 1.
+        }
+
+        // 8. Return seen.
+        Ok(seen.into_iter().collect::<Vec<JsString>>())
+    }
+
+    /// Returns an array containing the canonical locale names.
+    ///
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///  - [MDN docs][mdn]
+    ///
+    /// [spec]: https://tc39.es/ecma402/#sec-intl.getcanonicallocales
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/getCanonicalLocales
+    pub(crate) fn get_canonical_locales(
+        _: &JsValue,
+        args: &[JsValue],
+        context: &mut Context,
+    ) -> JsResult<JsValue> {
+        // 1. Let ll be ? CanonicalizeLocaleList(locales).
+        let ll = Self::canonicalize_locale_list(args, context)?;
+        // 2. Return CreateArrayFromList(ll).
+        Ok(JsValue::Object(Array::create_array_from_list(
+            ll.iter()
```

Comment:
```suggestion
            ll.into_iter()
```
Avoids the clone inside the map